### PR TITLE
Add LSP support and a simple first version only for vscode

### PR DESCRIPTION
**Pull request by ng-marcus**
This PR does quite a lot of prep work for the next version which will have a generic LSP server written in Reason.

For now, I wanted to get this first version out that will give people typechecking support whilst writing grain in vscode.

The repo is reorganised to move the vscode specific editor extension into a subfolder.    For this version, the client and server are only for vscode and so the server implementation is colocated with the client.  This will change with the next major version.

Building the extension package should be the same as before, by running "vsce package" in the editor-extensions/vscode folder.  This uses the suggested best practice of webpack to bundle the files in a production bundle.   This is my first vscode extension, so let me know if this doesn't work on a machine other than my own!

I've promoted the whole repo version to 0.5.0 - not sure if this is right or if we start lower for the reason-language-server project?  Also please check the package.json files for publisher, version etc to make sure this aligns with what's expected.

There are a couple of settings for the extension now, the key one being the enabled flag to let you stop using the LSP if it causes problems.  It's also assumed that the grain cli is on the user's path - this can be specified directly in the settings if not.

Any changes needed, please just let me know!

### CODE REVIEW

**Review by ospencer — COMMENTED** [review]
This looks really good! I have a lot of thoughts; let me know what you think!

Versioning: Do we really need the `package.json` at the root? Other than that one, I think it makes sense to have the client/server/extension all have separate versions. I'd be okay with setting the client version to just `0.0.0`, since there doesn't really seem like a real reason to keep it up to date/in sync (but let me know if I'm wrong).

Testing: I don't quite understand the testing setup 😅 
That's probably a matter of me not being super familiar with testing for vsc extensions, though. The first time I ran them I got some sort of error message, but the second time I didn't.

LSP: It works! I think when I was testing it out, it was hitting the compiler on every keystroke and was lagging a bit. I'm not sure if something was wrong with my local setup though. Either way, maybe we could debounce it? 

I'm really hyped to get this rolled out!

**Comment by ng-marcus** [comment]
Thanks for the feedback!   Yes, we can definitely lose the package.json at the root until there's a proper LSP server implementation there.   

For the testing, that's definitely something I can improve.

I just read up about debouncing for LSP which is what other implementations have done which is definitely worth getting in.  I'll have a go at that tomorrow.    One approach is to wait until there's a 350ms gap in typing and then process it, so I'll start there and see how it feels.  Were you testing on a single file, or did it have a lot of dependencies?

**Comment by ospencer** [comment]
Just one file! It seemed a little weird and that's totally possible with my local setup, haha.

**Comment by ng-marcus** [comment]
That's weird, it's pretty fast on my 2018 Macbook Pro but that's some decent hardware.  I'll try it on my Chromebooks too as it's always good to test on less performant hardware.   It would be calling the compiler on every keypress at the moment, so debouncing is a good thing to apply, especially for larger sets of files.

**Comment by ospencer** [comment]
I will say something is _probably_ up with my local setup. Programs typically compile instantly, so I didn't see how it would really be a problem.

**Comment by ospencer** [comment]
(I'm on a 2017 MBP if that's helpful.)

**Comment by ng-marcus** [comment]
I've updated the code to only call the compiler on changes at most once per second. That can be configured in the settings too (but requires a vscode restart at the moment to pick up).  Also, if you set the trace level to verbose you can see how often the compile process is being called in the grain language server output window.

For me the compile is pretty instant too, so it would be good to see if this denounced version helps.

**Comment by ng-marcus** [comment]
For the testing, it's pretty pointless at the moment, but the error you saw was probably because you can't run the tests which use a downloaded vscode whilst you have your own vscode open.

File: editor-extensions/vscode/client/src/test/completion.test.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+// 	test('Completes JS/TS in txt file', async () => {
+// 		await testCompletion(docUri, new vscode.Position(0, 0), {
+// 			items: [
+// 				{ label: 'JavaScript', kind: vscode.CompletionItemKind.Text },
+// 				{ label: 'TypeScript', kind: vscode.CompletionItemKind.Text }
+// 			]
+// 		});
+// 	});
+// });
```
Comment (ospencer) [inline editor-extensions/vscode/client/src/test/completion.test.ts:21]: Can probably just delete this, yeah?
```suggestion
```

File: editor-extensions/vscode/client/src/test/diagnostics.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// 	const docUri = getDocUri('diagnostics.txt');
+
+// 	test('Diagnoses uppercase texts', async () => {
+// 		await testDiagnostics(docUri, [
+// 			{ message: 'ANY is all uppercase.', range: toRange(0, 0, 0, 3), severity: vscode.DiagnosticSeverity.Warning, source: 'ex' },
+// 			{ message: 'ANY is all uppercase.', range: toRange(0, 14, 0, 17), severity: vscode.DiagnosticSeverity.Warning, source: 'ex' },
+// 			{ message: 'OS is all uppercase.', range: toRange(0, 18, 0, 20), severity: vscode.DiagnosticSeverity.Warning, source: 'ex' }
+// 		]);
+// 	});
+// });
```
Comment (ospencer) [inline editor-extensions/vscode/client/src/test/diagnostics.test.ts:20]:
```suggestion
```

File: editor-extensions/vscode/client/src/test/helper.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+export let editor: vscode.TextEditor;
+export let documentEol: string;
+export let platformEol: string;
+
+/**
+ * Activates the vscode.lsp-sample extension
+ */
+export async function activate(docUri: vscode.Uri) {
+	// The extensionId is `publisher.name` from package.json
+	const ext = vscode.extensions.getExtension('vscode-samples.lsp-sample')!;
```
Comment (ospencer) [inline editor-extensions/vscode/client/src/test/helper.ts:19]: Should we update this now, or are we just leaving this in until we've got tests?

File: editor-extensions/vscode/server/src/server.ts [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+// The content of a text document has changed. This event is emitted
+// when the text document first opened or when its content has changed.
+documents.onDidChangeContent(change => {
+
+	// Debounce the actual work as we don't want to run a compile on every keystroke
+
+	changedDocuments.add(change.document.uri);
+
+	//	validateWithCompiler(change.document.uri);
```
Comment (ospencer) [inline editor-extensions/vscode/server/src/server.ts:190]: And it looks like we can drop this now too?
```suggestion
```

File: editor-extensions/vscode/server/src/server.ts [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+	}
+
+
+	// Send the computed diagnostics to VSCode.
+	connection.sendDiagnostics({ uri: textDocumentUri, diagnostics });
+};
+
+connection.onDidChangeWatchedFiles(_change => {
+	// Monitored files have change in VSCode
+	connection.console.log('We received an file change event');
```
Comment (ospencer) [inline editor-extensions/vscode/server/src/server.ts:276]: Should we only do this in verbose mode?

**Comment by ng-marcus** [comment]
Thanks for the comments, fixing them right now!

File: editor-extensions/vscode/server/src/server.ts (hunk 5 above)
Comment (ng-marcus) [reply editor-extensions/vscode/server/src/server.ts:276]: Actually, we don't need this at all

File: editor-extensions/vscode/client/src/test/helper.ts (hunk 3 above)
Comment (ng-marcus) [reply editor-extensions/vscode/client/src/test/helper.ts:19]: Updated to reflect this package, but I'll leave this here.  There are no tests at the moment as this is such a simple wrapper, but they will be coming for the autocomplete and other work, and will be much more useful once this is re-implemented in Reason on the server side.

File: editor-extensions/vscode/client/src/test/diagnostics.test.ts (hunk 2 above)
Comment (ng-marcus) [reply editor-extensions/vscode/client/src/test/diagnostics.test.ts:20]: Yup, deleted

File: editor-extensions/vscode/client/src/test/completion.test.ts (hunk 1 above)
Comment (ng-marcus) [reply editor-extensions/vscode/client/src/test/completion.test.ts:21]: Yup, deleted

File: editor-extensions/vscode/server/src/server.ts (hunk 4 above)
Comment (ng-marcus) [reply editor-extensions/vscode/server/src/server.ts]: Cleaned up and deleted

**Review by ospencer — APPROVED** [review]
Absolutely amazing. I'll get this published ASAP!

**Pull request merged**